### PR TITLE
feat: Add mongodb as a provider

### LINKED ISSUE #624: Add mongodb as a provider
**Is your feature request related to a problem? Please describe.**
I'm looking to read data from mongodb in my workflow.

**Describe the solution you'd like**
In my workflow, i have few configurable parameters which are stored in mongodb. I want to fetch them against an evernt (like trigger interval, or email address etc) and use in workflow

**Describe alternatives you've considered**
as our configuraiton needs to be externalized, i'm not able to think of any other alternative. I'm open for discussion regarding this.

**Additional context**

### PR DESCRIPTION
**Pull request by rishi-raj-jain**
fix #624
/claim #624

### CODE REVIEW

**Comment by rishi-raj-jain** [comment]
@shahargl 

Ready for your review. Added both the URI and the Host, password method/

**Comment by shahargl** [comment]
@rishi-raj-jain cool! few things - 
1. Add Mongo png so that it will show on the platform
2. Add docs

**Comment by rishi-raj-jain** [comment]
@shahargl 

Done. Ready for your review.

File: keep/providers/mongodb_provider/mongodb_provider.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+import os
+from pymongo import MongoClient
+from keep.contextmanager.contextmanager import ContextManager
+from keep.providers.base.base_provider import BaseProvider
+from keep.providers.models.provider_config import ProviderConfig, ProviderScope
+import pydantic
+import dataclasses
+
+@pydantic.dataclasses.dataclass
```
Comment (b4s36t4) [inline keep/providers/mongodb_provider/mongodb_provider.py:9]: Can't we simplify? 

`from pydantic import dataclasses`

File: keep/providers/mongodb_provider/mongodb_provider.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    ) -> list | tuple:
+        """
+        Executes a query against the MongoDB database.
+
+        Returns:
+            list | tuple: list of results or single result if single_row is True
+        """
+        client = self.__generate_client()
+        database = client[self.authentication_config.database]
+        results = list(database.eval(query))
```
Comment (b4s36t4) [inline keep/providers/mongodb_provider/mongodb_provider.py:102]: https://www.mongodb.com/docs/v4.2/reference/method/db.eval/ Eval is not supported on latest mongodb versions?

File: keep/providers/mongodb_provider/mongodb_provider.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+        Returns:
+            pymongo.MongoClient: MongoDB Client
+        """
+        if self.authentication_config.uri:
+            client = MongoClient(self.authentication_config.uri)
+        else:
+            client = MongoClient(
+                f"mongodb://{self.authentication_config.username}:{self.authentication_config.password}@{self.authentication_config.host}/{self.authentication_config.database}"
+            )
```
Comment (b4s36t4) [inline keep/providers/mongodb_provider/mongodb_provider.py:74]: https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient can we follow parameter based implementation?

Comment (b4s36t4) [reply keep/providers/mongodb_provider/mongodb_provider.py:74]: or atleast escaping special characters using `urllib`?

**Comment by shahargl** [comment]
Nice! The mongo db icon should be square and without background (transparent) - e.g. https://github.com/keephq/keep/blob/main/keep-ui/public/icons/cloudwatch-icon.png

Fix it and I'll merge @rishi-raj-jain

File: keep/providers/mongodb_provider/mongodb_provider.py (hunk 3 above)
Comment (b4s36t4) [inline keep/providers/mongodb_provider/mongodb_provider.py:74]: What if the config values are empty (considering they are not required), can we handle the error case here?

File: keep/providers/mongodb_provider/mongodb_provider.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        """
+        Validates required configuration for MongoDB's provider.
+        """
+        self.authentication_config = MongoDBProviderAuthConfig(
+            **self.config.authentication
+        )
+
+    def _query(
+        self, query="", as_dict=False, single_row=False, **kwargs: dict
+    ) -> list | tuple:
```
Comment (b4s36t4) [inline keep/providers/mongodb_provider/mongodb_provider.py:93]: Are reason for kwargs?, if not can we remove it?

**Comment by b4s36t4** [comment]
Taking the liberty to review the PR, hope it's ok :)

**Comment by b4s36t4** [comment]
Screenshot 2023-12-12 at 4 05 00 PM

**Comment by rishi-raj-jain** [comment]
@shahargl 

Fixed the icon.

File: keep/providers/mongodb_provider/mongodb_provider.py (hunk 3 above)
Comment (rishi-raj-jain) [reply keep/providers/mongodb_provider/mongodb_provider.py:74]: Screenshot 2023-12-12 at 4 07 11 PM

The docs you linked anyways created the string.

Comment (b4s36t4) [reply keep/providers/mongodb_provider/mongodb_provider.py:74]: Yes, but they were escaped string, that's my 2nd point atleast making the values escaped or please see the available/acceptable paramters you can simply pass `host,port, username,password` as named parameters this way client will handle the malformed values.

**Comment by rishi-raj-jain** [comment]
> Screenshot 2023-12-12 at 4 05 00 PM

Oh looking into this!

**Comment by shahargl** [comment]
waiting for the `eval` resolving issue :)

**Comment by rishi-raj-jain** [comment]
@shahargl 

Fixed.

- example dict to be passed: https://www.mongodb.com/docs/manual/reference/command/find/#specify-a-sort-and-limit
- here's the function definition: https://pymongo.readthedocs.io/en/stable/api/pymongo/database.html#pymongo.database.Database.command

**Comment by shahargl** [comment]
I'll merge it and if @b4s36t4 have any other thoughts we will create a fix

**Comment by shahargl** [comment]
Can you fix the test run?
CleanShot 2023-12-12 at 13 32 38

**Comment by rishi-raj-jain** [comment]
Fixing...

**Comment by shahargl** [comment]
Looking at code, will it work? I don't think you define **query and **kwargs within the same function. why not just using query?

**Comment by rishi-raj-jain** [comment]
> Looking at code, will it work? I don't think you define **query and **kwargs within the same function. why not just using query?

Oh yes, I was trying out python spread operator.  Have changed.

**Comment by rishi-raj-jain** [comment]
fixed the query object example.

**Comment by rishi-raj-jain** [comment]
tests should be all good now. have followed the syntax as suggested in pymongo docs.

do let me know if anything.

**Comment by shahargl** [comment]
Merged!

**Pull request merged**